### ISSUE TITLE
Page with missing node in live workspace cannot be modified or deleted

### ISSUE THREAD

**Issue by niklasnatter** [sulu/sulu#5779]
| Q | A
| --- | ---
| Sulu Version | 2.1 (probably also 1.6)

# Actual Behavior

Sulu uses two PHPCR workspaces; a `draft` workspace for unpublished data and a `live` workspace for published data that is displayed on the website. The `draft` workspace can be inspected via ` bin/adminconsole doctrine:phpcr:shell` and the `live` workspace can be inspected via `bin/websiteconsole doctrine:phpcr:shell`. If everything runs smooth, both workspaces always contain a node for each page of the application. Even if a page was never published, the `live` workspace will contain a node that looks like this:

```
+-----------------+-------------+--------------------------------------+
| jcr:mixinTypes  | NAME (17)   | [0] mix:referenceable                |
| jcr:primaryType | NAME (15)   | nt:unstructured                      |
| jcr:uuid        | STRING (36) | 405266e0-97ca-44dd-9808-990ad9d4a022 |
+-----------------+-------------+--------------------------------------+
```

This node is created by the `PublishSubscriber::createNodeInPublicWorkspace` method when the page is created. Unfortunately, it is possible that an error happens while creating the page and we will **end up in a state where the node is present in the `draft` workspace but not created in the `live` workspace**. In this scenario, the page will be visible in the administration interface, but every operation (modify, delete) will lead to an exception like these:

```
Next PHPCR\PathNotFoundException: Item /cmf/example/contents/page-title not found in workspace default_live in /sulu-skeleton/vendor/jackalope/jackalope/src/Jackalope/Session.php:289
Stack trace:
#0 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/DocumentManagerBundle/Session/Session.php(84): Jackalope\Session->getNode('/cmf/example/co...', -1)
#1 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Document/Subscriber/PublishSubscriber.php(267): Sulu\Bundle\DocumentManagerBundle\Session\Session->getNode('/cmf/example/co...')
#2 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Document/Subscriber/PublishSubscriber.php(101): Sulu\Bundle\PageBundle\Document\Subscriber\PublishSubscriber->getLiveNode(Object(Sulu\Bundle\PageBundle\Document\PageDocument))
#3 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(270): Sulu\Bundle\PageBundle\Document\Subscriber\PublishSubscriber->createNodeInPublicWorkspace(Object(Sulu\Component\DocumentManager\Event\PersistEvent), 'sulu_document_m...', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#4 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(230): Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}(Object(Sulu\Component\DocumentManager\Event\PersistEvent), 'sulu_document_m...', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#5 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(59): Symfony\Component\EventDispatcher\EventDispatcher->callListeners(Array, 'sulu_document_m...', Object(Sulu\Component\DocumentManager\Event\PersistEvent))
#6 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/DocumentManager/DocumentManager.php(57): Symfony\Component\EventDispatcher\EventDispatcher->dispatch(Object(Sulu\Component\DocumentManager\Event\PersistEvent), 'sulu_document_m...')
#7 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Controller/PageController.php(577): Sulu\Component\DocumentManager\DocumentManager->persist(Object(Sulu\Bundle\PageBundle\Document\PageDocument), 'en', Array)
#8 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Controller/PageController.php(350): Sulu\Bundle\PageBundle\Controller\PageController->persistDocument(Object(Symfony\Component\HttpFoundation\Request), 'Sulu\\Bundle\\Pag...', Object(Sulu\Bundle\PageBundle\Document\PageDocument), 'en')
#9 /sulu-skeleton/vendor/symfony/http-kernel/HttpKernel.php(157): Sulu\Bundle\PageBundle\Controller\PageController->putAction(Object(Symfony\Component\HttpFoundation\Request), '64da17ce-6036-4...')
#10 /sulu-skeleton/vendor/symfony/http-kernel/HttpKernel.php(79): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#11 /sulu-skeleton/vendor/symfony/http-kernel/Kernel.php(195): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#12 /sulu-skeleton/public/index.php(66): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#13 {main}
```

```
Next PHPCR\PathNotFoundException: Item /cmf/example/contents/page-title not found in workspace default_live in /sulu-skeleton/vendor/jackalope/jackalope/src/Jackalope/Session.php:289
Stack trace:
#0 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/DocumentManagerBundle/Session/Session.php(84): Jackalope\Session->getNode('/cmf/example/co...', -1)
#1 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/Content/Document/Subscriber/StructureRemoveSubscriber.php(116): Sulu\Bundle\DocumentManagerBundle\Session\Session->getNode('/cmf/example/co...')
#2 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/Content/Document/Subscriber/StructureRemoveSubscriber.php(94): Sulu\Component\Content\Document\Subscriber\StructureRemoveSubscriber->removeReferences(Object(Sulu\Bundle\PageBundle\Document\PageDocument))
#3 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/Content/Document/Subscriber/StructureRemoveSubscriber.php(82): Sulu\Component\Content\Document\Subscriber\StructureRemoveSubscriber->removeDocument(Object(Sulu\Bundle\PageBundle\Document\PageDocument))
#4 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(270): Sulu\Component\Content\Document\Subscriber\StructureRemoveSubscriber->handleRemove(Object(Sulu\Component\DocumentManager\Event\RemoveEvent), 'sulu_document_m...', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#5 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(230): Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}(Object(Sulu\Component\DocumentManager\Event\RemoveEvent), 'sulu_document_m...', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#6 /sulu-skeleton/vendor/symfony/event-dispatcher/EventDispatcher.php(59): Symfony\Component\EventDispatcher\EventDispatcher->callListeners(Array, 'sulu_document_m...', Object(Sulu\Component\DocumentManager\Event\RemoveEvent))
#7 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/DocumentManager/DocumentManager.php(63): Symfony\Component\EventDispatcher\EventDispatcher->dispatch(Object(Sulu\Component\DocumentManager\Event\RemoveEvent), 'sulu_document_m...')
#8 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Controller/PageController.php(406): Sulu\Component\DocumentManager\DocumentManager->remove(Object(Sulu\Bundle\PageBundle\Document\PageDocument))
#9 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Component/Rest/RestControllerTrait.php(146): Sulu\Bundle\PageBundle\Controller\PageController->Sulu\Bundle\PageBundle\Controller\{closure}('64da17ce-6036-4...')
#10 /sulu-skeleton/vendor/sulu/sulu/src/Sulu/Bundle/PageBundle/Controller/PageController.php(411): Sulu\Component\Rest\AbstractRestController->responseDelete('64da17ce-6036-4...', Object(Closure))
#11 /sulu-skeleton/vendor/symfony/http-kernel/HttpKernel.php(157): Sulu\Bundle\PageBundle\Controller\PageController->deleteAction(Object(Symfony\Component\HttpFoundation\Request), '64da17ce-6036-4...')
#12 /sulu-skeleton/vendor/symfony/http-kernel/HttpKernel.php(79): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#13 /sulu-skeleton/vendor/symfony/http-kernel/Kernel.php(195): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#14 /sulu-skeleton/public/index.php(66): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#15 {main}
```

# Workaround

At the moment, if this happens in an application, **the affected node must be deleted from the `draft` workspace manually**. This is possible via the PHPCR shell that can be started with `bin/websiteconsole doctrine:phpcr:shell` or via the `bin/websiteconsole doctrine:phpcr:node:remove` command.  

# Expected Behavior

I think it should at least be possible to remove or republish the affected page via the administration interface.

# Steps to Reproduce

1. Create a page
2. Remove the node of the page from the `live` workspace via `bin/websiteconsole doctrine:phpcr:shell`  or `bin/websiteconsole doctrine:phpcr:node:remove`
3. Try to modify the page in the administration interface

# Possible Solutions

It might be enough to adjust the `PublishSubscriber` to automatically create the node in the `live`workspace like shown below. This would create the missing node when an affected page is modified. I am not sure if this has any undesired side effects. I think **we would at least need a similar change in the `StructureRemoveSubscriber`**.

```diff
     public function createNodeInPublicWorkspace(PersistEvent $event)
     {
         $node = $event->getNode();
 
-        if ($node->isNew()) {
+        try {
+            $liveNode = $this->getLiveNode($event->getDocument());
+        } catch (PathNotFoundException $e) {
             $this->createNodesWithUuid($node);
 
             return;
         }
 
-        $liveNode = $this->getLiveNode($event->getDocument());
         $nodeName = $node->getName();
 
         if ($liveNode->getName() !== $nodeName) {
             $liveNode->rename($nodeName);
         }
     }
```

**Comment by themuddfamily** [comment]
The solution worked perfect for me +1 to this :)

**Comment by alexander-schranz** [comment]
@nnatter I think we could optimize the function as we know that if `$node->isNew()` the `getLiveNode` always will fail. So we could go there directly call `createNodesWithUuid` without try to load the livenode.

**Comment by themuddfamily** [comment]
Seems like this breaks creation of new pages - returning
"Uncaught PHP Exception PHPCR\PathNotFoundException" "Item X not found in workspace default_live"
I suspect this is partly what @alexander-schranz is referring to so I changed the fix to this.

I realise there is probably a way better way to do this but I wanted to post a way that works for others.

```diff
     public function createNodeInPublicWorkspace(PersistEvent $event)
     {
         $node = $event->getNode();

+        if($node->isNew()){
+          $this->createNodesWithUuid($node);
+
+          return;
+        }
 

-        if ($node->isNew()) {
+        try {
+            $liveNode = $this->getLiveNode($event->getDocument());
+        } catch (PathNotFoundException $e) {
             $this->createNodesWithUuid($node);
 
             return;
         }
 
-        $liveNode = $this->getLiveNode($event->getDocument());
         $nodeName = $node->getName();
 
         if ($liveNode->getName() !== $nodeName) {
             $liveNode->rename($nodeName);
         }
     }
```

**Comment by alexander-schranz** [comment]
@themuddfamily thx for the response. I hope we can implement such a fix to avoid this kind of problems in the future.

**Comment by niklasnatter** [comment]
This was fixed in #6174 and will be included in the 2.2.12 and 2.3.2 release 🙂  

In the new releases, the missing node is automatically created when the corrupt page is modified. If you want to remove a corrupt page, you need to modify the page a single time. After this, the deletion should work without any errors.
